File: 백준/Gold/17485. 진우의 달 여행 （Large）/진우의 달 여행 （Large）.js
const input = require("fs")
  .readFileSync("/dev/stdin")
  .toString()
  .trim()
  .split("\n")

const [n, m] = input[0].split(" ").map(Number)
const values = [Array.from({ length: m }, () => 0)].concat(
  input.slice(1).map((row) => row.split(" ").map(Number)),
)

const dp = Array.from({ length: n + 1 }, () =>
  Array.from({ length: m }, () => Array.from({ length: 3 }, () => Infinity)),
)

for (let i = 0; i < m; i++) {
  for (let j = 0; j < 3; j++) {
    dp[0][i][j] = 0
  }
}
const dir = [
  [1, 1],
  [1, 0],
  [1, -1],
]

for (let r = 1; r <= n; r++) {
  for (let c = 0; c < m; c++) {
    for (let direction = 0; direction < 3; direction++) {
      const [pr, pc] = [r - dir[direction][0], c - dir[direction][1]]

      if (pr < 0 || pc < 0 || pr > n || pc >= m) continue
      if (c === 0 && direction === 0) {
        continue
      } else if (c === m - 1 && direction === 2) {
        continue
      } else {
        dp[r][c][direction] =
          values[r][c] +
          Math.min(
            dp[pr][pc][(direction + 1) % 3],
            dp[pr][pc][(direction + 2) % 3],
          )
      }
    }
  }
}

console.log(Math.min(... dp[n].flat()))

File: 백준/Gold/17485. 진우의 달 여행 （Large）/진우의 달 여행 （Large）.py
import sys
input = sys.stdin.readline

# 입력 처리
n, m = map(int, input().strip().split())
values = [[0] * m] + [list(map(int, input().strip().split())) for _ in range(n)]

# DP 테이블 초기화 (n+1 x m+2 x 3 크기)
dp = [[[1000000] * 3 for _ in range(m + 2)] for _ in range(n + 1)]

# 초기 조건
for i in range(1, m + 1):
    for j in range(3):
        dp[0][i][j] = 0

# DP 계산
for r in range(1, n + 1):
    for c in range(1, m + 1):
        dp[r][c][0] = values[r][c - 1] + min(dp[r - 1][c - 1][1], dp[r - 1][c - 1][2])
        dp[r][c][1] = values[r][c - 1] + min(dp[r - 1][c][0], dp[r - 1][c][2])
        dp[r][c][2] = values[r][c - 1] + min(dp[r - 1][c + 1][0], dp[r - 1][c + 1][1])

# 결과 출력
print(min(min(dp[n][c]) for c in range(1, m + 1)))
